Add garbage collector

mokagio opened the pull request:
This PR adds a Ruby script that:

- Checks the versions on Bintray for a give list of libraries
- Filters out the dev version: those that got built from a PR
- For each dev version, checks if the PR is closed
- If the PR is closed, it deletes the version

The script will run every day at midnight UTC via CircleCI.

This PR is already quite long because of the test. I extracted the setup code in #5, to make it easier to review.

I'll merge it only after #5 has been merged.

## A note on the Ruby structure

There's likely a better and more Ruby-y way of structuring the code, perhaps by building a gem.

I went for this slightly hacked setup because of the scope of this project.

I'm definitely keen on refining it if necessary, but since it's isolated and it works™ I think this is a good place where to leave it for now (unless there's a clear mistake on my side on how it's organized, of course)

## Testing

It's a bit hard to manually test this code, that's the reason I invested in unit tests. **Please have a look at the tests and let me know if they are clear**.

To test it live, I opened a few PRs on my test repo and changed the schedule to run every five minutes (see diff in #7).

The [first run](https://app.circleci.com/pipelines/github/Automattic/bintray-garbage-collector/89/workflows/840b6d87-d9c7-4760-a114-89e536e9fd48/jobs/76) deleted the versions for the closed PR:

![image](https://user-images.githubusercontent.com/1218433/97557355-c7553500-1a2e-11eb-8a0a-d1e692d94e50.png)

The [second](https://app.circleci.com/pipelines/github/Automattic/bintray-garbage-collector/90/workflows/783a4a7b-f383-4cff-839a-edd7e2dd167b/jobs/77) didn't find anything to delete.

![image](https://user-images.githubusercontent.com/1218433/97557870-772aa280-1a2f-11eb-99c4-2c2184da2132.png)

mokagio committed:
Add object interfacing with the Bintray APIs

mokagio committed:
Add object interfacing with the GitHub APIs

Decided not to use a gem because we need to do only one API call.

mokagio committed:
Run unit tests in CI

mokagio committed:
Add object coordinating the garbage collection

mokagio committed:
Add script to call to run garbage collection

mokagio committed:
Run garbage collection from CI nightly

mokagio commented on bin/bintray_garbage_collector.rb:
```diff
@@ -0,0 +1,25 @@
+require 'dotenv'
+require './lib/garbage_collector.rb'
+
+# We might want to extra this in a file to make it easier to edit
+projects = [
+  { bintray: 'wordpress-mobile/maven/utils', github: 'wordpress-mobile/WordPress-Utils-Android' }
+]
```
I'll do the extraction in a file when it'll be time to add another package to the list.

AliSoftware commented on bin/bintray_garbage_collector.rb:
```diff
@@ -0,0 +1,23 @@
+require 'dotenv'
+require './lib/garbage_collector.rb'
+
+# We might want to extra this in a file to make it easier to edit
```
```suggestion
# We might want to extract this in a file to make it easier to edit
```

AliSoftware commented on bin/bintray_garbage_collector.rb:
```diff
@@ -0,0 +1,23 @@
+require 'dotenv'
+require './lib/garbage_collector.rb'
```
```suggestion
require_relative '../lib/garbage_collector.rb'
```

otherwise it will try to reference relative to the current pwd (as opposed to the path relative to the current file)

mokagio replied:
Nice catch! Thanks

AliSoftware commented on .circleci/config.yml:
```diff
@@ -1,14 +1,27 @@
+    executor: ruby
+    steps:
+      - checkout
+      - run: bundle install && bundle exec ruby ./bin/bintray_garbage_collector.rb
```
How meta, using a ruby tool (bundler) to run (potentially another) ruby itself 😅 

(I think that's how you're supposed to do it indeed when you have `ruby 'xx'` in your `Gemfile` like you did, so that seems correct to me, but threw me off a bit at first 😅 )

AliSoftware commented on bin/bintray_garbage_collector.rb:
```diff
@@ -0,0 +1,23 @@
+  { bintray: 'wordpress-mobile/maven/utils', github: 'wordpress-mobile/WordPress-Utils-Android' }
+]
+
+Dotenv.load unless ENV['CI']
```
nit: should we use `unless ENV['CI']` here or should we prefer to test if the dotfile if it exists (which it should not on CI, only the `.env.sample` would) – which feels maybe like a more logical condition for deciding to load it or not?

mokagio replied:
Very good point. While this script is not really meant to run by devs, leaking the fact that we might run it in CI in the implementation is not nice.

AliSoftware commented on lib/garbage_collector.rb:
```diff
@@ -0,0 +1,60 @@
+
+class GarbageCollector
+
+  def initialize(bintray_user, bintray_key, verbose = false, dry_run = false)
```
nit/suggestion: now that Ruby allows it I like to use named arguments whenever possible (especially for the ones with default values as they can be omitted) to make APIs clearer, e.g.:
```suggestion
  def initialize(bintray_user, bintray_key, verbose: false, dry_run: false)
```

Then at call site it would look like:
```
GarbageCollector.new(bintray_user, bintray_key, verbose: verbose, dry_run: dry_run).run(projects)
```

But it will also allow us to pass `dry_run: true` without having to explicitly pass a value (and thus using the default) for `verbose` (which wouldn't be possible without named args)

nbd tho, just thought I'd mention it in case you didn't know about this feature but feel free to keep as is if you prefer.

mokagio replied:
Thanks for bringing this up. I vaguely knew about this feature, but never used it. It makes total sense and I love how it makes up (partly) the lack of compiler to help know what parameter to pass where.

> But it will also allow us to pass dry_run: true without having to explicitly pass a value (and thus using the default) for verbose (which wouldn't be possible without named args)

This is even more so a good reason to use it. Although, in the particular case of how this object is used in the `bin/` script, I think we'll always want to explicitly pass both values. I would have liked to make them options (`--dry-run`, `--verbose`), but run out of time. I might do that in a followup PR when I'll add more projects to the list.

AliSoftware commented on lib/garbage_collector.rb:
```diff
@@ -0,0 +1,60 @@
+        if @github.is_pr_closed?(project[:github], pr_id)
+          log "Deleting versions for PR #{pr_id}"
+        else
+          log "PR #{pr_id} is not closed. Skipping."
```
nit 😛 
```suggestion
          log "PR #{pr_id} is still open. Skipping."
```

AliSoftware replied:
PS: Also super extra nitpick, but could be nice having a simple visual way (some prefix char? or use ANSI colors? 😛 ) to easily differentiate at first glance log messages that are announces ("Checking xxx..." – adding "..." to indicate it might take the time of an HTTP request to do it) vs the ones that are actions/replies.

For example:
```
Looking for versions to delete for *******/maven/utils-experiment...
> Found versions for PR(s): 3, 2, 40
Checking if versions for PR 3 should be deleted...
> PR 3 is not closed. Skipping.
Checking if versions for PR 2 should be deleted...
> Deleting versions for PR 2...
> Deleted version 2-d869dc6
Checking if versions for PR 40 should be deleted...
> PR 40 is not closed. Skipping.
```

_That's my OCD speaking so feel free to ignore_ 😅

mokagio replied:
Thanks for the suggestions @Automattic/jetpack-design 

- Active positive language FTW 👍 
- I like the prefix char idea (I love colors, too, but might be a bit of overhead for this tool, at least right now) I thought about using `\t` originally, but left it out because I didn't like the look of it. This is much better!

AliSoftware commented on lib/garbage_collector.rb:
```diff
@@ -0,0 +1,60 @@
+
+        versions.each do |version|
+          if @dry_run
+            log "DRY RUN! Deleted version #{version}"
```
```suggestion
            log "DRY RUN! Would have deleted version #{version}"
```

mokagio replied:
Nice. Yeah, this wording makes more sense.

AliSoftware commented on lib/github_client.rb:
```diff
@@ -0,0 +1,24 @@
+require 'net/http'
+require 'json'
+
+class GitHubClient
```
Why not use the `octokit` gem provided by GitHub? 🤔  Not worth it for so little API surface we need I guess? (sounds a reasonable rationale so fine with me to keep it that way, but just mentioning in case you didn't know about octokit)

mokagio replied:
Yep, that's exactly it (as per 1e17a70's description).

If we'll need to do more advanced requests or processing with GitHub in the future, `octokit` would be my choice.

AliSoftware commented on lib/bintray_client.rb:
```diff
@@ -0,0 +1,53 @@
+
+  def get_bintray_versions(project)
+    uri = URI("#{@base_url}/#{project}")
+    http = Net::HTTP.new(uri.host, uri.port)
+    http.use_ssl = true
+    request = Net::HTTP::Get.new(uri.request_uri)
+
+    response = http.request(request)
+
+    return nil unless response.kind_of? Net::HTTPSuccess
```
Do you think it could be worth extracting those lines in a small helper method, given that the same is used for the `GitHubClient` (and a very similar one is used for `delete_bintray_version` except for the `Net::HTTP::Delete`)?

mokagio replied:
Love a good DRY refactor suggestion. Thanks!

AliSoftware commented on lib/bintray_client.rb:
```diff
@@ -0,0 +1,53 @@
+
+    return versions
+      .select { |v| v =~ /^\d+-.*$/ }
+      .group_by { |v| v.split('-').first }
```
So typical Bintray versions naming convention we use is `<prnum>-<xxx>` is that right? (might be worth a comment?)

mokagio replied:
Good point. Totally worth a comment.

AliSoftware commented on lib/bintray_client.rb:
```diff
@@ -0,0 +1,53 @@
+
+    return nil if message.nil?
+
+    return message == 'success'
```
Are you telling me that BinTray is those kind of APIs that can return `200 OK` but still return an error in the http body? 🤦

AliSoftware commented on lib/bintray_client.rb:
```diff
@@ -0,0 +1,53 @@
+
+class BintrayClient
+
+  def initialize(user, key, base_url = 'https://bintray.com/api/v1/packages')
```
ditto, opinionated but could be nice to use a named param for `base_url`

AliSoftware commented on spec/bintray_client_spec.rb:
```diff
@@ -1,3 +1,131 @@
+      it 'returns false' do
+        # See https://www.jfrog.com/confluence/display/BT/Bintray+REST+API#BintrayRESTAPI-DeleteVersion
+        stub_request(:delete, "#{base_url}/#{package}/versions/#{version}")
+          .to_return(status: 200, body: { message: 'some error' }.to_json)
```
the infamous HTTP "200 error"… 😭 🤦

mokagio replied:
Yep... Not ideal 😞

AliSoftware commented on spec/github_client_spec.rb:
```diff
@@ -1,3 +1,72 @@
+      end
+    end
+
+    context 'when the response has not state key' do
```
```suggestion
    context 'when the response has no state key' do
```

AliSoftware commented on lib/bintray_client.rb:
```diff
@@ -0,0 +1,53 @@
+    @key = key
+  end
+
+  def get_bintray_versions(project)
```
Would be nice to have rubydoc comments for all the methods of all classes in `lib/`

AliSoftware reviewed (commented):
Looks good!

Mainly a couple of nits and small fixes, but also I think you should add rubydoc documentation to the methods, especially to declare the expected `@param`s and `@return` values/types, to help maintenance and code reading (given that ruby isn't a typed language)

mokagio committed:
Use require_relative in bin script

mokagio committed:
Remove ENV["CI"] check from script

The `Dotenv.load` call doesn't fail if there is not file and since we
have a check for the ENV values to be available, I opted for omitting a
check for the file before attempting to load it.

mokagio committed:
Use active positive language in logs (not closed -> open)

mokagio committed:
Add "..." suffix and ">" prefix to make log output clearer

Rationale at https://github.com/Automattic/bintray-garbage-collector/pull/6#discussion_r515287061

Thanks to @AliSoftware for the suggestion!

mokagio committed:
Use "Would have deleted" instead of "deleted" in dry run log

Thanks @AliSoftware.

See
https://github.com/Automattic/bintray-garbage-collector/pull/6#discussion_r515287367.

mokagio committed:
Extract logic to make HTTPS calls returning JSON in object

Thanks @AliSoftware

See https://github.com/Automattic/bintray-garbage-collector/pull/6#discussion_r515288779

mokagio committed:
Extract the failure context from the success one in BintrayClient spec

mokagio committed:
Use keyword arguments for clearer code

Thanks for the great suggestion @AliSoftware

See discussion at
https://github.com/Automattic/bintray-garbage-collector/pull/6#discussion_r515286165

loremattei commented on .circleci/config.yml:
```diff
@@ -1,14 +1,27 @@
+    executor: ruby
     steps:
-      - run: echo "Hello, world! 👋"
+      - checkout
```
Well... I guess this is a good example of "show the history of your project since the very beginning" ;-)

loremattei reviewed (approved):
Code looks good to me and the tests make sense. 
I have only two suggestions:
- I think we should make the tests required to merge a PR in the configuration of this repository because it looks like we're going to rely on them quite a lot to validate changes to this project. 

- I think we should add a Slack notification about the result of the job. Probably, we want to have notifications on failures, but since this is a new scheduled job, we may also have a notification on successful runs for the firsts few weeks.

AliSoftware commented on lib/bintray_client.rb:
```diff
@@ -0,0 +1,40 @@
+
+  def initialize(user:, key:, base_url: 'https://bintray.com/api/v1/packages')
+    @base_url = base_url
+    @user = user
+    @key = key
+    @http_client = HTTPClient.new
```
Shouldn't you pass the user/key to the `HTTPClient.new`, since it's using it in its impl (see below)?
Or if you keep it at `BintrayClient` level (which might make more sense), you'll need to pass them as parameters to `delete_json`

AliSoftware commented on lib/http_client.rb:
```diff
@@ -0,0 +1,30 @@
+    http = Net::HTTP.new(uri.host, uri.port)
+    http.use_ssl = true
+    request = Net::HTTP::Delete.new(uri.request_uri)
+    request.basic_auth @user, @key
```
I don't see those properties defined in `HTTPClient`? Those are defined in `BintrayClient` (but `BintrayClient` doesn't inherit `HTTPClient` anyway and uses composition instead – which I think was a good choice – so those are not gonna be available here).

You should probably pass `user` and `key` as parameters to `delete_json` instead – probably as an optional hash which only calls `basic_auth` if that hash is defined, to make it even more flexible

```suggestion
    request.basic_auth auth[:user], auth[:key] unless auth.nil?
```

---

Once you get that in, you could even consider going further with the refactoring, something like this perhaps?

```ruby
  # @param [Net::HTTP class] req_klass
  #        The Net::HTTP class to use to create the request, e.g. Net::HTTP::Get or Net::HTTP::Delete
  # @param [URI] uri The URI to sent the request to
  # @param [Hash] auth If not nil, should be a hash with keys `:user` and `key` that will be used for Basic Auth
  # @return [Hash|Array] The parsed JSON response, or nil if the request was not a HTTPSuccess.
  #
  def json_request(req_klass, uri:, auth: nil)
    http = Net::HTTP.new(uri.host, uri.port)
    http.use_ssl = (uri.scheme == 'https')
    request = req_klass.new(uri.request_uri)
    request.basic_auth auth[:user], auth[:key] unless auth.nil?

    response = http.request(request)

    return nil unless response.kind_of? Net::HTTPSuccess

    JSON.parse(response.body)
  end

  def get_json(uri:, auth: nil)
    json_request(Net::HTTP::Get, uri: uri, auth: auth)
  end

  def delete_json(uri:, auth: nil)
    json_request(Net::HTTP::Delete, uri: uri, auth: auth)
  end
```

mokagio replied:
Great catch! I thought my unit tests were comprehensive and only run the refactored code in dry run mode. I had overlooked this.

I now have a `.with(basic_auth: ['user', 'key'])` expectation in the test to delete the package which fails with the version of the code on which this comment was made.

I like the DRY suggestion to extract the class, I was looking for something like that but hadn't found it yet.

AliSoftware reviewed (changes requested):
Found an issue with some of your member variables not on the right class that will need to be fixed first. Rest lgtm.

(Also, don't forget to add doc comments to methods 😉 )

mokagio committed:
Rename HTTPClient to JSONHTTPClient

The aim is to make it clearer that this is an HTTP client to make
requests that expect a JSON response.

mokagio committed:
Pass valid HTTP auth parameters to delete call

Thank you for catching this @AliSoftware!

See https://github.com/Automattic/bintray-garbage-collector/pull/6#discussion_r518636934

mokagio committed:
Extract req generation and JSON response parsing in JSONHTTPClient

Thanks for the suggestion @AliSoftware.

See https://github.com/Automattic/bintray-garbage-collector/pull/6#discussion_r518636934

mokagio committed:
Add documentation for types in lib/ using YARD

https://yardoc.org/

I chose yard because it seems to be the de-facto standard.

mokagio committed:
Add comment regarding Bintray dev versions format

Thanks @AliSoftware.

See
https://github.com/Automattic/bintray-garbage-collector/pull/6#discussion_r515289841

mokagio commented on lib/bintray_client.rb:
```diff
@@ -0,0 +1,67 @@
+  ##
+  # A new instance of +BintrayClient+
+  #
+  # @param user [String] the username to use to authenticate with the Bintray API
```
I copied this formatting from the example in the [YARD documentation for `@param`](https://rubydoc.info/gems/yard/file/docs/Tags.md#param)

![image](https://user-images.githubusercontent.com/1218433/98500813-8f0cec80-22a1-11eb-8a1d-058521e8e5a9.png)

mokagio replied:
Also, some of the examples [in the docs](https://rubydoc.info/gems/yard/file/docs/GettingStarted.md) have a new line between the method description and the list of tags, while others don't. I decided to always add one because I seemed more readable like that.

mokagio commented on lib/garbage_collector.rb:
```diff
@@ -0,0 +1,73 @@
+  # Run the garbage collection on the given array of projects
+  #
+  # @param projects [Array] the array of +Hash+ projects on which to run the garbage collection
+  def run(projects)
```
I thought about making this an object as well, the same way `HTTPBasicAuth` is. I decided to save it for when I'll extract the projects list configuration in a dedicated file, which I plan to do when we'll need to add a new project to the list.

mokagio commented on lib/json_http_client.rb:
```diff
@@ -0,0 +1,46 @@
+
+##
+# +JSONHTTPClient+ makes HTTP request to endpoint that should respond with a JSON and returns the parsed response
+class JSONHTTPClient
```
Renamed from `HTTPClient` to `JSONHTTPClient`. It's more verbose but makes it clearer that this is an opinionated layer (HTTP requests with JSON response) and simplifies the API to `get` and `delete` which I thought was neater.

Finally, the `webmock` gem we use [depends](https://github.com/bblimke/webmock/blob/474c6ab2a0fb933982a2cc4f5fe8b66acc9ce7b6/webmock.gemspec#L40) on an [`HTTPClient`](https://github.com/nahi/httpclient) already so this keeps things separated.

mokagio commented:
Thank you for your re-review @AliSoftware and for catching that nasty issue with the DELETE with no auth. (😳)

I hadn't forgotten about the docs just hadn't finished with them. I should have reverted my PR to draft, to make it clear it was still WIP and save you some time.

Now it's ready, review re-requested. Thank you very much.

AliSoftware commented on .env.sample:
```diff
@@ -0,0 +1,2 @@
+BINTRAY_USER=
+BINTRAY_KEY=
```
We should add a `README` with instructions about those. I mean in practice those will likely be provided by CI rather than run locally, but still.

AliSoftware commented on Gemfile:
```diff
@@ -6,6 +6,10 @@ ruby File.read('.ruby-version').strip
 
 gem "dotenv"
 
+group :dev do
+  gem "yard"
+end
```
👍

AliSoftware commented on lib/bintray_client.rb:
```diff
@@ -0,0 +1,63 @@
+  #
+  # @param project [String] the name of the project on Bintray
+  # @param version [String] the version to delete from +project+
+  # @return [Bool] whether the deletion succeeded or +nil+ if the request failed
```
Is it worth having a `Bool?` here? Might make it simpler to return `false` if the request failed and avoid `nil` here?

AliSoftware commented on lib/github_client.rb:
```diff
@@ -0,0 +1,30 @@
+  #
+  # @param repo [String] the name of the GitHub repository with the "org/name" format, e.g. "Automattic/bintray-garbage-collector"
+  # @param pr_id [Integer] the pull request id
+  # @return [Bool] whether the pull request is closed or nil if the request fails
```
ditto, wondering if we shouldn't just return `false` if request fails to avoid optional bool?

AliSoftware commented on lib/http_basic_auth.rb:
```diff
@@ -0,0 +1,16 @@
+# Encapuslates the parameters to use for basic HTTP authentication
+#
+# @see https://docs.ruby-lang.org/en/2.7.0/Net/HTTP.html#class-Net::HTTP-label-Basic+Authentication
+class HTTPBasicAuth
+  attr_reader :user
+  attr_reader :password
+
+  # A new instance of +HTTPBasicAuth+
+  #
+  # @param user [String] the username
+  # @param password [String] the password
+  def initialize(user:, password:)
+    @user = user
+    @password = password
+  end
+end
```
Tip: You could also just use ruby's `Struct` to create data classes like that in a more concise way (could be useful too for when you decide to convert your `projects` into one too) with a oneliner like that:

```
HTTPBasicAuth = Struct.new(:user, :password, keyword_init: true)
```

See https://www.rubyguides.com/2017/06/ruby-struct-and-openstruct/

Not sure if you find it more readable or more obscure… 🤷

AliSoftware reviewed (approved):
Left some comments but it's just ideas that I think they could as well go into a separate PR on top of this one for future improvements

mokagio commented:
Hey @AliSoftware. I wanted to address your suggestions before merging because they resonate, but I got distracted with another project and I'm afraid I'll run out before going on leave for the next two weeks.

I promise I'll address your comments once back in December.
